Mark the corresponding UserLevel when a PeerReview is submitted

joshlory opened the pull request:
This change adds the correct progress dot colors for each state of a `peer_reviewable` level.

<img src="https://cloud.githubusercontent.com/assets/413693/15484702/20acc52e-20f0-11e6-844c-af32e567ed86.png" width="550">

The `submitted` flag is separate from `best_result`.  Submitting an attempt sets `best_result` to -50 ("attempted").  Un-submitting an accepted or rejected answer also sets `best_result` back to -50.

The highest `best_result` the client is allowed to cache locally is 999.  Higher values can't be cached by the client because they may cycle up and down depending on actions outside the client's control, such as another using submitting a peer review.
- [x] TODO: Add unit tests.

mehalshah commented on code-studio/src/js/clientState.js:
```diff
@@ -89,17 +89,20 @@ clientState.levelProgress = function (scriptName, levelId) {
   a = a || 0;
   b = b || 0;
-  if (a === 0) {
+  if (a === 0 || a > 100) {
```
Is it 100 because of the constant here -https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/constants.js ? Can we reference that instead of the magic number?

mehalshah commented on code-studio/src/js/progress.js:
```diff
@@ -121,18 +121,26 @@ progress.renderStageProgress = function (stageData, progressData, clientProgress
     var status;
-    if (serverProgress && serverProgress[level.id] && serverProgress[level.id].submitted) {
+    var result = (serverProgress[level.id] || {}).result;
+    if (serverProgress && serverProgress[level.id] && serverProgress[level.id].reviewable && result == 150 || result == 200) {
```
Similar magic number concerns

mehalshah commented on dashboard/app/models/peer_review.rb:
```diff
@@ -30,14 +30,53 @@ class PeerReview < ActiveRecord::Base
   REVIEWS_PER_SUBMISSION = 2
+  REVIEWS_FOR_CONSENSUS = 2
 
   enum status: {
```
I should have mentioned this earlier, but convention with PLC objects has been to use strings as statuses instead of numbers. It makes for much easier debugging, and we have space to burn.

joshlory replied:
@philbogle recommended the Rails built-in enum type.  It seems like a good fit for this use case.

mehalshah commented on dashboard/app/models/peer_review.rb:
```diff
@@ -30,14 +30,53 @@ class PeerReview < ActiveRecord::Base
 
+  def mark_user_level
+    user_level = UserLevel.find_by(user: submitter, level: level)
+
```
Handle nil here - I would raise a specific exception since I would expect this to never happen.

Or just do find_by!

joshlory replied:
Maybe we should add a FK to `UserLevel` and make it `null: false`?  (If we always expect this to be present.)

mehalshah replied:
Do UserLevels get deleted?

joshlory replied:
@laurelfan would know for sure.  In practice, I don't think they are deleted.

laurelfan replied:
they don't get deleted in the normal course of operations. I wouldn't preclude having some one-off reason to delete user levels though (CS request, migration...).

joshlory replied:
That makes sense, I'll update to `find_by!`.

mehalshah commented on lib/cdo/activity_constants.rb:
```diff
@@ -10,4 +10,7 @@ module ActivityConstants
   FREE_PLAY_RESULT = 30
   BEST_PASS_RESULT = 100
+
+  REJECTED_RESULT = 150
```
These constants are specific to peer review and should have that incorporated in their name

joshlory replied:
Good call, fixed in 3a911a98d7d1de8eb852ca015f995e3a395e5f6e and b7213572f9d666acca361d8e31cd3096c64140c8.

mehalshah commented on shared/css/user-progress.scss:
```diff
@@ -60,6 +60,16 @@ a.level_link.submitted:hover
   color: white !important;
 }
 
+.level_link.accepted, .level_link.accepted:link, .level_link.accepted:visited {
```
Not for this PR, but I'm bothered by how many places we have white !important. Excessively using !important is a sign of poorly organized stylesheet hierarchy. Also it sounds racist.

joshlory replied:
Agreed that we use `!important` too much :D

joshlory commented:
@mehalshah I believe this is ready for another pass.  Sorry the scope of this PR has grown so much!

mehalshah commented on code-studio/src/js/progress.js:
```diff
@@ -56,9 +64,19 @@ progress.populateProgress = function (scriptName, puzzlePage) {
-      // Only the server can speak to whether a level is submitted.  If it is,
-      // we show the submitted styling.
-      if (serverProgress[levelId].submitted) {
+      // Only the server can speak to whether a level is submitted/accepted/rejected.  If it is,
```
During HOC when we are in high scale mode, what will this look like? Will the call to get progress for these dots be allowed?

joshlory replied:
They will be yellow ("in progress") because that is the highest level of progress the client is allowed to write down locally.

In reality these level types should never be included in a high-scale script because it doesn't make sense to write `FreeResponse` answers and submit them unless there is a server present.

mehalshah commented on dashboard/app/helpers/application_helper.rb:
```diff
@@ -48,9 +48,15 @@ def check_mark_html
-    if submitted
+    result = user_level.try(:best_result)
+
+    if result == Activity::REVIEW_REJECTED_RESULT
```
Convention I'm used to (though I don't know what the performance impact is) is to put the order of these in the most common normal case to least common normal case. So submitted would be on top

joshlory replied:
A level can be both `submitted` and have `result == Activity::REVIEW_REJECTED_RESULT`.

mehalshah replied:
Can a level have state rejected/accepted without submitted? If not, then maybe it should be

```
if user_level.try(:submitted)
    if (user_level.try(:rejected))
       rejected
    elseif (user_level.try(:accepted))
        accepted
    else
        submitted
else
   'not_tried'
```

mehalshah replied:
I'm annoyed that github took away all my indenting in the above comment

laurelfan replied:
it's still there, I added a ``` 

(if you didn't know already, github lets other people edit your comments)

joshlory replied:
Yes, `rejected`/`accepted` are independent of `submitted`.  `submitted` has lower priority.

You can use ``` (three backticks) to format code and it will preserve indentation.

mehalshah commented on dashboard/app/models/peer_review.rb:
```diff
@@ -30,14 +30,53 @@ class PeerReview < ActiveRecord::Base
+      script: script,
+      level: level,
+      level_source_id: most_recent
+    ).where.not(status: nil)
```
If all you care about is the # of reviews and count, you could just pluck(:status).compact! (or do a select uniq)

joshlory replied:
I actually want the full review object, since plucking `:status` will only give me the raw enum value (0, 1 or 2).

joshlory replied:
Hmm but you're right that I can re-write the next lines to use the enum scopes instead of `.all?`.

mehalshah commented on dashboard/app/models/peer_review.rb:
```diff
@@ -30,14 +30,53 @@ class PeerReview < ActiveRecord::Base
+    elsif reviews.all?(&:rejected?)
+      user_level.update!(best_result: Activity::REVIEW_REJECTED_RESULT)
+    else
+      # TODO: find_or_create PeerReview assigned to the instructor
```
Will this be a later PR (i don't mind if it is - this one has grown large)

joshlory replied:
Yes.  Good catch, I'll track this in the doc.

mehalshah commented on dashboard/app/models/peer_review.rb:
```diff
@@ -46,15 +85,26 @@ def localized_status_description
+      script: user_level.script,
+      level: user_level.level,
+    ).where.not(id: created).destroy_all
   end
```
Any reason you don't do this first?

Also, consider wrapping this in a transaction?

begin transaction
-remove old reviews
-make new reviews
end transaction

joshlory replied:
Good catch, fixed in de8ea9f3220ff8d462c4f204823327abbac5c397.

mehalshah commented on dashboard/test/models/peer_review_test.rb:
```diff
@@ -22,7 +22,64 @@ class PeerReviewTest < ActiveSupport::TestCase
     initial = PeerReview.count
-    @user.track_level_progress_async(script_level: @script_level, new_result: 100, submitted: true, level_source_id: level_source.id)
+    @user.track_level_progress_async(script_level: @script_level, new_result: Activity::UNSUBMITTED_RESULT, submitted: true, level_source_id: level_source.id)
     assert_equal PeerReview::REVIEWS_PER_SUBMISSION, PeerReview.count - initial
```
Is this guaranteed to work? The message gets enqueued in the progress queue, but is there a guarantee that it gets dequeued before the assert statement?

You might want to try track_level_progress_sync instead

joshlory replied:
I's "guaranteed" in the sense that these tests are run in normal (not high-scale) mode.  I can update them to use `track_level_progress_sync`.

joshlory replied:
Changed in de8ea9f3220ff8d462c4f204823327abbac5c397.

mehalshah commented:
Looked through - fewer comments this time, some concerns about how this acts during HOC

joshlory commented:
Addressed your feedback — in high-scale mode this feature will degrade gracefully, but the expectation is that Peer Review isn't used in high-scale scripts.

mehalshah commented:
LGTM

ryansloan commented:
LGTM
